BosunReporter/Handlers/DataDogStatsdMetricHandler.cs

```diff
@@ -0,0 +1,294 @@
+
+            switch (payloadType)
+            {
+                case PayloadType.Counter:
+                case PayloadType.CumulativeCounter:
+                case PayloadType.Gauge:
+                    return SendMetricAsync(payloadType, buffer);
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(payloadType));
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void SerializeMetadata(IBufferWriter<byte> writer, IEnumerable<MetaData> metadata)
+        {
+            // this particular implementation doesn't understand metadata
+        }
+
+        /// <inheritdoc />
+        protected override void SerializeMetric(IBufferWriter<byte> writer, in MetricReading reading)
+        {
+            // UTF-8 formatted as follows:
+            // {metric}:{value}|{unit}|{tag},{tag}
+
+            // exit early if we don't support the metric
+            byte[] unit;
+            switch (reading.Type)
+            {
+                case MetricType.Counter:
+                case MetricType.CumulativeCounter:
+                    unit = s_counter;
+                    break;
+                case MetricType.Gauge:
+                    unit = s_gauge;
+                    break;
+                default:
+                    // not supported yet!
+                    throw new NotSupportedException("Metric type '" + reading.Type + "' is not supported");
+            }
+
+            // calculate the length of buffer we need
+            var encoding = Encoding.UTF8;
+            var length = encoding.GetByteCount(reading.Name);
+            if (!string.IsNullOrEmpty(reading.Suffix))
+            {
+                length += encoding.GetByteCount(reading.Suffix);
+            }
+
+            length += s_pipe.Length;
+
+            var valueLength = 1 + 1 + ValueDecimals; // first number + decimal point + number of decimals
+            var value = reading.Value;
+            var valueAsInt = (int)value;
+            while ((valueAsInt /= 10) > 0)
+            {
+                valueLength++;
+            }
+
+            length += valueLength;
+            length += s_pipe.Length + unit.Length;
+
+            if (reading.Tags.Count > 0)
+            {
+                length += s_pipe.Length + s_hash.Length;
+                foreach (var tag in reading.Tags)
+                {
+                    length += encoding.GetByteCount(tag.Key) + s_colon.Length + encoding.GetByteCount(tag.Value) + s_comma.Length;
+                }
+
+                // take away the last comma
+                length -= s_comma.Length;
+            }
+
+            length += s_newLine.Length;
+
+            // go grab some space in the buffer writer
+            var memory = writer.GetMemory(length);
+            // we also need a byte array otherwise we can't do all the encoding bits
+            if (!MemoryMarshal.TryGetArray(Unsafe.AsRef<ReadOnlyMemory<byte>>(memory), out var arraySegment))
```

Comment:
I would advise *against* this; if your usage here is because of `Encoding`, then either target ns2.1 (which has APIs for this), or use `unsafe` (you can use `fixed` with a `Span<byte>` to get a `byte*` for use with `Encoding`). Either way, I'd move that code into a helper methor